docs/cupynumeric/source/user/Legate Task tutorial.rst

```diff
@@ -0,0 +1,981 @@
+
+Main function
+--------------
+Let’s take a look at the input and output parameters for this SAXPY example.
+
+.. code-block:: python
+
+    size = args.size
+    
+    x_global = cpn.arange(size, dtype=cpn.float32)
+    y_global = cpn.ones(size, dtype=cpn.float32)
+    z_global = cpn.zeros(size, dtype=cpn.float32)
+      
+    start = time()
+    saxpy_task(x_global, y_global, z_global, 2.0)
+    end = time()
+    
+    print(f"\nTime elapsed for saxpy: {(end - start)/1000:.6f} milliseconds")
+
+For this example, three one-dimensional arrays of default size 1000 are
+created. ``x_global`` contains values from 0 to 999, ``y_global`` is filled with
+ones, and ``z_global`` is initialized with zeros to store the result. The
+saxpy_task function is then called to compute the operation ``z_global = 2.0 * x_global + y_global``. We can change the size of the arrays
+through the ``--size`` command-line argument when running the script.
+
+Task function
+-------------
+The following example shows how to define a task function that performs the SAXPY operation.
+
+.. code-block:: python
+
+    @task(variants = (VariantCode.CPU, VariantCode.GPU,),
+          constraints = (align("x", "y"),
+                         align("y", "z")))
+    def saxpy_task(ctx: TaskContext, x: InputArray, y: InputArray, z: OutputArray, a: float) -> None:
+       xp = cupy if ctx.get_variant_kind() == VariantCode.GPU else numpy
+       x_local = xp.asarray(x)
+       y_local = xp.asarray(y)
+       z_local = xp.asarray(z)
+       z_local[:] = a * x_local + y_local
+
+The constraint used is ``align``, it is used to ensure that ``x``, ``y`` , and ``z``
+are partitioned in the same way. This is so that corresponding elements
+live together on the same device. For example, imagine there are 4 GPUs,
+and the problem size is 1000.
+
+- GPU 1 gets the range 0–249
+
+- GPU 2 gets the range 250–499
+
+- GPU 3 gets the range 500–749
+
+- GPU 4 gets the range 750–999
+
+With the usage of ``align(“x”, “y”)`` and ``align(“y”, “z”)`` constraints, we
+make sure that ``x[i]``, ``y[i]``, and ``z[i]`` are all assigned to the same gpu. If
+we want to compute ``z[2]``, and GPU 1 handles the calculation for it, ``x[2]``
+and ``y[2]`` need to be handled in the same GPU in order to get the correct
+answer. Given the ``align`` constraint, Legate will handle co-location of
+corresponding elements across arrays, ensuring correctness.
+
+The ``saxpy_task`` function uses ``TaskContext`` and its ``get_variant_kind()``
+method to determine the execution target (GPU or CPU) and accordingly
+create views of the task-local data as NumPy or CuPy arrays. It then performs the SAXPY operation element-wise by computing
+``z_local[:] = a * x_local + y_local``. This task runs in parallel on the
+available hardware (CPU or GPU), enabling efficient computation.
+
+Complete module
+---------------
+Putting the pieces above together, here is a complete module that
+can be run with the ``legate`` command line launcher:
```

Comment:
Instead of putting this code inline, you should make a standalone file `examples/saxpy.py`, and include the file verbatim here. For the code snippets above, you should also do the same. This will ensure things are properly formatted and tested, and don't inadvertently bitrot.